src/request_queue.js

```diff
@@ -240,165 +268,225 @@ export class RequestQueue {
+
+        // 1) Queue head index is ahead of the main table and the request is not present in the table yet.
+        //    In this case, keep the request marked as in progress for a short while,
+        //    so that isFinished() doesn't return true and _ensureHeadIsNonEmpty() doesn't not load the request
+        //    into the queueHeadDict straight again. After the interval expires, fetchNextRequest()
+        //    will try to fetch this request again, until it eventually appears in the main table.
+        if (!request) {
+            log.debug('Cannot find request from the beginning of queue, will try again later', { nextRequestId });
+            setTimeout(() => {
+                delete this.inProgress[nextRequestId];
+            }, STORAGE_CONSISTENCY_DELAY_MILLIS);
+            return null;
+        }
+
+        // 2) Queue head index is behind the main table and the underlying request was already handled
+        //    (by some other client, since we keep the track of handled requests in recentlyHandled dictionary).
+        //    We just add the request to the recentlyHandled dictionary so that next call to _ensureHeadIsNonEmpty()
+        //    will not put the request again to queueHeadDict.
+        if (request.handledAt) {
+            log.debug('Request fetched from the beginning of queue was already handled', { nextRequestId });
+            this.recentlyHandled.add(nextRequestId, true);
+            return null;
+        }
+
+        return request;
     }
 
     /**
-     * Marks request handled after successful processing.
+     * Marks request as handled after successful processing.
+     * Handled requests will never again be returned by the {@link RequestQueue#fetchNextRequest} function.
      *
      * @param {Request} request
      * @return {Promise<QueueOperationInfo>}
      */
-    markRequestHandled(request) {
+    async markRequestHandled(request) {
         // TODO: This function should also support object instead of Apify.Request()
         validateMarkRequestHandledParams(request);
 
-        if (!this.requestIdsInProgress[request.id]) {
-            throw new Error(`Cannot mark request ${request.id} as handled as it is not in progress!`);
+        if (!this.inProgress[request.id]) {
+            throw new Error(`Cannot mark request ${request.id} as handled, because it is not in progress!`);
         }
 
         if (!request.handledAt) request.handledAt = new Date();
 
-        return requestQueues
-            .updateRequest({
-                request,
-                queueId: this.queueId,
-                clientKey: this.clientKey,
-            })
-            .then((queueOperationInfo) => {
-                this._removeFromInProgress(request.id);
-                this._cacheRequest(getRequestId(request.uniqueKey), queueOperationInfo);
+        const queueOperationInfo = await requestQueues.updateRequest({
```

Comment:
I know this is old, but it's really confusing, because the `queueOperationInfo` returned from API is not the same `QueueOperationInfo` we have in SDK and down below, the assigment of `queueOperationInfo.request = request` is just a mindfuck.

Reply:
Yeah, this is a good point. I would completely remove the request object from this. It's not really needed.